### DISCUSSION TITLE
Mouse Navigation

### DISCUSSION THREAD

**Discussion by ClaudeJGreengrass** [ATMakersOrg/hid-remapper#3] [Ideas]
Mouse navigation of modern Graphical User Interfaces (GUI) is in my opinion a 90/10 affair. 90% of the time, precision is not required.  Just move the pointer from where in the direction of the next goal be it an icon, a file, a browser tab, or any other “hotspot” that is clickable in the GUI.  The other 10% is the close in precision movement requiring fine motor control.  Needless to say, a skill that is generally lacking in people with tremors in their hands.

Since we have two parts to the navigation of the GUI, I suggest we have two separate configurations.  The normal/fast/big movement configuration with a default to light filtering and non attenuation of movements and a second, close up, fine movement configuration with a default to stronger filtering and some attenuation of the mouse movement.  Both of these configuration options are user settable.

Attenuation is used to further reduce the pointer's response to mouse movements.  A 2:1 attenuation would require a 2 pixel movement to achieve a 1 pixel response.  Fractional attenuation might be possible.  It is an area to be explored. 

How do we decide if we should be using configuration 1 or configuration 2?  I suggest the speed of the mouse.  I don’t know what the common units of movement are.  When the mouse movements slow down we switch from the fast/general configuration to the slow/precision configuration.  Switching from the slow to the fast configuration is dependent on a fast continuous movement. The timing threshold to be determined.

**Comment by dglaude** [comment]
The kind of "mode" you are talking about looks a lot like the DPI (dot per inch) sensibility configuration of advanced mouse. I think on some mouse you have additional button, either on the side or up and down of the middle roller to change the sensibility of the mouse.

I don't know if those button are common and what usb-hid code is used for that, but if it is standard and can be captured by hid-remapper kind of software, the that would be a good candidate to activate mode or to adjust a parameter from the algorithm. But that would only be limited to a subset of mouse.

Right now I have been using the middle button as a toggle between mode. I use it for turning on/off the modification I want to make, just like in the example "expressions: middle button enables mouse jiggler". And for DPI switching, there is the "middle button cycles DPI".

What is missing in hid-remapper to have a parameter that can be used in expression, is a kind of "variable" witch value that can be changed (like by the middle button wheel). We do have the scaling value that can be changed with web interface, but it is not the same. The example to cycles DPI is using one layer per possible DPI value to have factor 1, 1/2 or 1/4, but this is a lot of mapping duplication and we are limited by the number of layers available.

**Reply by ClaudeJGreengrass** [reply]
Using one of the "other buttons", not the common left and right could easily be used to change the "mode".  It could even be accumulative.  Fast Speed then click Other -> Medium speed click Other again -> Slow speed.  Clicking Other again reverts to Fast Speed.  This is a sound proposal. Easy to to understand and easy to implement in code.  We need to ask some end users if it  is a good fit for their needs.

As to expressions, please open up a separate thread for discussion of variables and stateful data and to the need of more layers.  Perhaps configuration variables/data should be accessible from expressions.
